pandas/core/indexes/numeric.py

```diff
@@ -225,7 +240,13 @@ def astype(self, dtype, copy=True):
                 # TODO(jreback); this can change once we have an EA Index type
                 # GH 13149
                 arr = astype_nansafe(self._values, dtype=dtype)
-                return Int64Index(arr, name=self.name)
+                if isinstance(self, Float64Index):
+                    return Int64Index(arr, name=self.name)
+                else:
+                    return NumericIndex(arr, name=self.name, dtype=dtype)
+        elif self._is_numeric_index:
+            if not is_extension_array_dtype(dtype) and is_numeric_dtype(dtype):
```

Comment:
why is this needed?

Reply:
The super method is `Index.astype`. That method needs to assume that numeric indexes are `Int64Index` etc. and not `NumericIndex` in order to maintain backward compatability.

So without the lines above e.g. `NumericIndex[int32].astype("int8")` would return a `Int64Index` with int64 dtype , which is not what we'd want.

Reply:
kk can you leave a comment to this effect